File: src/utils/code-scanner.ts
/**
 * Code Scanner — NFR Static Analysis
 *
 * Scans source files for hardcoded secrets, passwords, API keys,
 * connection strings, and other security violations.
 */

import { promises as fs } from "fs";
import path from "path";
import chalk from "chalk";
import { performAiVulnerabilityScan } from "./llm";

export interface ScanFinding {
  ruleId: string;
  category: string;
  description: string;
  severity: "ERROR" | "WARNING" | "INFO";
  filePath: string;
  lineNumber: number;
  matchedContent: string;
  remediation: string;
}

export interface ScanReport {
  scannedFiles: number;
  totalFindings: number;
  errors: number;
  warnings: number;
  infos: number;
  findings: ScanFinding[];
  scannedAt: string;
}

interface ScanRule {
  id: string;
  description: string;
  severity: "ERROR" | "WARNING" | "INFO";
  pattern: RegExp;
  remediation: string;
}

/**
 * @deprecated These hardcoded regex rules are retained as a deterministic 
 * baseline / fallback scanner. In a production environment, use actual 
 * tools like Semgrep or Gitleaks for primary scanning, followed by AI analysis.
 */
const SCAN_RULES: ScanRule[] = [
  {
    id: "SEC-001",
    description: "Hardcoded password detected",
    severity: "ERROR",
    pattern: /(?:password|passwd|pwd)\s*[:=]\s*["'][^"']{3,}["']/gi,
    remediation:
      "Move password to Vault. Use: secretProvider.getSecret('DB_PASSWORD')",
  },
  {
    id: "SEC-002",
    description: "Hardcoded API key detected",
    severity: "ERROR",
    pattern:
      /(?:api[_-]?key|apikey|api[_-]?secret)\s*[:=]\s*["'][^"']{8,}["']/gi,
    remediation: "Move API key to Vault or environment variable.",
  },
  {
    id: "SEC-003",
    description: "Hardcoded connection string detected",
    severity: "ERROR",
    pattern: /(?:mongodb|postgres|mysql|mssql|redis):\/\/[^\s"']+/gi,
    remediation: "Use parameterized config from config-manager.ts.",
  },
  {
    id: "SEC-004",
    description: "Hardcoded IP address (non-localhost)",
    severity: "WARNING",
    pattern: /\b(?!127\.0\.0\.1|0\.0\.0\.0)(?:\d{1,3}\.){3}\d{1,3}\b/g,
    remediation: "Move IP address to config/environments/*.json.",
  },
  {
    id: "SEC-005",
    description: "Inline process.env usage without config abstraction",
    severity: "INFO",
    pattern: /process\.env\.[A-Z_]{3,}/g,
    remediation:
      "Use loadConfig/getConfig instead of direct process.env access.",
  },
  {
    id: "SEC-006",
    description: "Hardcoded external URL",
    severity: "INFO",
    pattern:
      /https?:\/\/(?!localhost|127\.0\.0\.1)[a-zA-Z0-9][a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g,
    remediation: "Move external URLs to config/environments/*.json.",
  },
  {
    id: "SEC-007",
    description: "Private key material detected",
    severity: "ERROR",
    pattern: /-----BEGIN\s+(?:RSA\s+|EC\s+)?PRIVATE\s+KEY-----/gi,
    remediation: "NEVER store private keys in source code. Use Vault.",
  },
  {
    id: "SEC-008",
    description: "AWS credential pattern detected",
    severity: "ERROR",
    pattern: /AKIA[0-9A-Z]{16}/g,
    remediation: "Use IAM roles or Vault for AWS credentials.",
  },
  {
    id: "SEC-009",
    description: "Bearer token pattern detected",
    severity: "ERROR",
    pattern: /["']Bearer\s+[A-Za-z0-9\-._~+/]+=*["']/g,
    remediation: "Move bearer tokens to Vault or environment variables.",
  },
  {
    id: "SEC-010",
    description: "Base64-encoded secret pattern",
    severity: "WARNING",
    pattern:
      /(?:secret|token|credential)\s*[:=]\s*["'][A-Za-z0-9+/]{20,}={0,2}["']/gi,
    remediation: "Verify this is not a hardcoded secret. Move to Vault.",
  },
];

const SCANNABLE_EXT = [".ts", ".js", ".json", ".env", ".yaml", ".yml"];
const SKIP_DIRS = [
  "node_modules",
  ".git",
  "dist",
  ".sdlc",
  "coverage",
  "tests",
];
const SKIP_FILES_FOR_RULES: Record<string, string[]> = {
  "SEC-005": [
    "env.ts",
    "vault.ts",
    "config-manager.ts",
    "config-schema.ts",
    "server.ts",
  ],
  "SEC-006": [
    "base.json",
    "dev.json",
    "staging.json",
    "prod.json",
    "Jenkinsfile",
    "package.json",
    "package-lock.json",
  ],
};

export async function runCodeScan(
  scanDir: string,
  filesToScan?: string[],
): Promise<ScanReport> {
  const files = filesToScan || (await collectFiles(scanDir));
  const findings: ScanFinding[] = [];

  for (const filePath of files) {
    const content = await fs.readFile(filePath, "utf8");
    
    // PERFORM AI VULNERABILITY SCAN (REPLACES REGEX RULES)
    const aiFindings = await performAiVulnerabilityScan(filePath, content);
    const lines = content.split(/\r?\n/);

    for (const f of aiFindings) {
      let correctedLineNumber = f.lineNumber || 0;

      // AI often miscounts line numbers. We verify and auto-correct here.
      if (f.matchedContent && f.matchedContent !== "N/A") {
        const expectedLine = lines[correctedLineNumber - 1] || "";
        if (!expectedLine.includes(f.matchedContent)) {
          // The reported line is wrong. Search for the actual content in the file.
          const actualIndex = lines.findIndex((l) =>
            l.includes(f.matchedContent),
          );
          if (actualIndex !== -1) {
            correctedLineNumber = actualIndex + 1;
          }
        }
      }

      findings.push({
        ruleId: f.ruleId || "AI-GENERIC",
        category: f.category || "General Security",
        description: f.description || "No description provided",
        severity: f.severity || "INFO",
        filePath: filePath.replace(/\\/g, "/"),
        lineNumber: correctedLineNumber,
        matchedContent: f.matchedContent || "N/A",
        remediation: f.remediation || "Review findings with a security expert.",
      });
    }
  }

  return {
    scannedFiles: files.length,
    totalFindings: findings.length,
    errors: findings.filter((f) => f.severity === "ERROR").length,
    warnings: findings.filter((f) => f.severity === "WARNING").length,
    infos: findings.filter((f) => f.severity === "INFO").length,
    findings,
    scannedAt: new Date().toISOString(),
  };
}

export function formatScanReport(report: ScanReport): string[] {
  const lines: string[] = [
    chalk.bold.underline("Code Security Scan Results"),
    `Scanned: ${chalk.cyan(report.scannedFiles)} files | ${chalk.gray(report.scannedAt)}`,
    `Findings: ${chalk.bold(report.totalFindings)} total — ` +
    `${chalk.bold.red(report.errors + " ERROR")}, ` +
    `${chalk.bold.yellow(report.warnings + " WARNING")}, ` +
    `${chalk.bold.cyan(report.infos + " INFO")}`,
    "",
  ];

  if (report.findings.length === 0) {
    lines.push(chalk.bold.green("✓ No security violations detected."));
    return lines;
  }

  // Get unique categories and sort them
  const categories = Array.from(
    new Set(report.findings.map((f) => f.category)),
  ).sort();

  for (const cat of categories) {
    const items = report.findings.filter((f) => f.category === cat);
    if (items.length === 0) continue;

    lines.push(chalk.bold.blue(`── ${cat.toUpperCase()} ${"─".repeat(Math.max(0, 44 - cat.length))}`));
    for (const f of items) {
      const shortPath = f.filePath.split("/").slice(-3).join("/");
      let severityIcon = "○";
      let severityColor = chalk.white;

      if (f.severity === "ERROR") {
        severityIcon = "✗";
        severityColor = chalk.bold.red;
      } else if (f.severity === "WARNING") {
        severityIcon = "⚠";
        severityColor = chalk.bold.yellow;
      } else if (f.severity === "INFO") {
        severityIcon = "ℹ";
        severityColor = chalk.bold.cyan;
      }

      lines.push(`  [${chalk.gray(f.ruleId)}] ${severityColor(severityIcon)} ${severityColor(f.description)}`);
      lines.push(`    File: ${chalk.cyan(shortPath)}:${chalk.yellow(f.lineNumber)}`);
      lines.push(`    Match: ${chalk.red(f.matchedContent)}`);
      lines.push(`    Fix: ${chalk.green(f.remediation)}`);
      lines.push("");
    }
  }

  lines.push(chalk.italic.gray("Remediation: Address findings based on severity and category risk."));
  return lines;
}

export async function saveScanLog(report: ScanReport): Promise<string> {
  const logDir = path.join(process.cwd(), ".sdlc", "logs");
  try {
    await fs.mkdir(logDir, { recursive: true });
    const now = new Date();
    const date = now.toISOString().split("T")[0].replace(/-/g, "");
    const time = now.toTimeString().split(" ")[0].replace(/:/g, "");
    const logFile = path.join(logDir, `nexus-security-report_${date}_${time}.json`);
    await fs.writeFile(logFile, JSON.stringify(report, null, 2));
    return logFile;
  } catch (err) {
    console.error("Failed to save scan log:", err);
    return "";
  }
}

function redactMatch(content: string): string {
  if (content.length <= 10) return content.slice(0, 3) + "***";
  return content.slice(0, 4) + "****" + content.slice(-2);
}

async function collectFiles(dir: string): Promise<string[]> {
  const files: string[] = [];
  try {
    const entries = await fs.readdir(dir, { withFileTypes: true });
    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      if (entry.isDirectory()) {
        if (!SKIP_DIRS.includes(entry.name))
          files.push(...(await collectFiles(fullPath)));
        continue;
      }
      const ext = path.extname(entry.name).toLowerCase();
      if (
        SCANNABLE_EXT.includes(ext) ||
        entry.name === ".env" ||
        entry.name === ".env.example"
      ) {
        if (!entry.name.includes("package-lock.json")) {
          files.push(fullPath);
        }
      }
    }
  } catch {
    /* skip unreadable dirs */
  }
  return files;
}

File: src/utils/code_scanner.py
"""Code Scanner — replaces code-scanner.ts. AI-driven SAST with rule-based fallback."""
from __future__ import annotations
import re
from pathlib import Path
from ..core.types import ScanFinding, ScanReport
from ..utils.llm import perform_ai_vulnerability_scan
from ..utils.scan_history import (
    snapshot_files, detect_changes, save_scan_result,
    load_last_scan, load_previous_findings, FileFingerprint
)

_RULES = [
    ("SEC-001", "Hardcoded Secret", r"(?:password|secret|api[_-]?key)\s*[:=]\s*['\"][^'\"]+['\"]", "ERROR", "Credentials", "Use environment variables or a vault."),
    ("SEC-002", "Hardcoded Token", r"(?:token|bearer|jwt)\s*[:=]\s*['\"][A-Za-z0-9+/=]{16,}['\"]", "ERROR", "Credentials", "Rotate and use vault."),
    ("SEC-003", "eval() Usage", r"\beval\s*\(", "ERROR", "Injection", "Remove eval() calls."),
    ("SEC-004", "exec() Usage", r"\bexec\s*\(", "WARNING", "Injection", "Avoid exec(); use subprocess."),
    ("SEC-005", "SQL Injection", r"(?:execute|query)\s*\(\s*['\"].*\+.*['\"]|f['\"].*(?:SELECT|INSERT|UPDATE|DELETE)", "ERROR", "Injection", "Use parameterized queries."),
    ("SEC-006", "Insecure HTTP", r"http://(?!localhost|127\\.0\\.0\\.1)", "WARNING", "Transport", "Use HTTPS."),
    ("SEC-007", "AWS Key Pattern", r"AKIA[0-9A-Z]{16}", "ERROR", "Credentials", "Rotate immediately; use IAM roles."),
    ("SEC-008", "Private Key", r"-----BEGIN (?:RSA |EC )?PRIVATE KEY-----", "ERROR", "Credentials", "Never commit private keys."),
    ("SEC-009", "Connection String", r"(?:mongodb|postgres|mysql|redis)://[^\\s]+:[^\\s]+@", "ERROR", "Credentials", "Use vault."),
    ("SEC-010", "Console Logging", r"\bprint\s*\(.*(?:password|secret|token|key)", "WARNING", "Logging", "Never log sensitive data."),
    ("SEC-011", "Debug Mode", r"DEBUG\s*=\s*True|debug\s*:\s*true", "WARNING", "Configuration", "Disable debug in production."),
    ("SEC-012", "CORS Wildcard", r"(?:allow_origins|Access-Control-Allow-Origin).*\\*", "WARNING", "Configuration", "Restrict CORS origins."),
    ("SEC-013", "Weak Crypto", r"\b(?:md5|sha1)\s*\(", "WARNING", "Cryptography", "Use SHA-256 or bcrypt."),
    ("SEC-014", "Path Traversal", r"\.\./|\.\.\\\\", "WARNING", "Injection", "Validate and sanitize paths."),
    ("SEC-015", "Disabled SSL", r"verify\s*=\s*False|rejectUnauthorized\s*:\s*false", "WARNING", "Transport", "Enable SSL verification."),
]
_EXTENSIONS = {".py",".js",".ts",".jsx",".tsx",".java",".go",".rs",".rb",".php",".cs",".json",".yaml",".yml",".env",".tf",".sh"}
_SKIP_DIRS = {"node_modules",".git","dist","build","__pycache__",".venv","venv",".sdlc"}

async def run_code_scan(scan_dir: str | Path, file_filter: list[Path] | None = None) -> ScanReport:
    scan_path = Path(scan_dir)
    findings: list[ScanFinding] = []; scanned = 0
    files_to_scan = file_filter if file_filter else list(_iter_files(scan_path))
    for file_path in files_to_scan:
        scanned += 1
        try: content = file_path.read_text(encoding="utf-8", errors="ignore")
        except Exception: continue
        rel = str(file_path.relative_to(scan_path))
        # Rule-based scan
        for rule_id, desc, pattern, sev, cat, fix in _RULES:
            for i, line in enumerate(content.splitlines(), 1):
                if re.search(pattern, line, re.IGNORECASE):
                    findings.append(ScanFinding(rule_id, cat, desc, sev, rel, i, line.strip()[:80], fix))
        # AI deep scan for .py/.ts/.js files
        if file_path.suffix in (".py", ".ts", ".js") and len(content) < 8000:
            ai_findings = await perform_ai_vulnerability_scan(rel, content)
            for f in ai_findings:
                if isinstance(f, dict) and "ruleId" in f:
                    findings.append(ScanFinding(f.get("ruleId","AI"), f.get("category","AI"), f.get("description",""),
                        f.get("severity","WARNING"), rel, f.get("lineNumber",0), f.get("matchedContent","")[:80], f.get("remediation","")))
    
    # Run platform-aware PySA (with Bandit fallback) scan
    try:
        from ..utils.pysa_scanner import run_pysa_scan
        pysa_findings = await run_pysa_scan(scan_path)
        findings.extend(pysa_findings)
    except Exception:
        pass

    errors = sum(1 for f in findings if f.severity == "ERROR")
    warnings = sum(1 for f in findings if f.severity == "WARNING")
    info = sum(1 for f in findings if f.severity == "INFO")
    return ScanReport(scanned, len(findings), findings, errors, warnings, info)

async def run_incremental_scan(scan_dir: str | Path) -> tuple[ScanReport, list[str], Path]:
    """Run an incremental scan: only re-scan files that changed since the last scan.
    
    Returns (report, formatted_output, scan_log_dir).
    Uses SHA-256 file fingerprinting for change detection — no git dependency.
    """
    scan_path = Path(scan_dir)
    all_files = list(_iter_files(scan_path))

    # Build current fingerprint snapshot
    current_fps = snapshot_files(all_files, scan_path)

    # Load previous scan
    prev_data, prev_fps = load_last_scan()

    delta_info = None
    if prev_data and prev_fps:
        # Detect changes
        delta = detect_changes(current_fps, prev_fps)
        changed_files = set(delta["changed"] + delta["new"])
        deleted_files = set(delta["deleted"])

        if not changed_files and not deleted_files:
            # Nothing changed — carry forward entire previous report
            prev_findings = load_previous_findings(prev_data)
            report = ScanReport(
                scanned_files=len(all_files),
                total_findings=len(prev_findings),
                findings=prev_findings,
                errors=sum(1 for f in prev_findings if f.severity == "ERROR"),
                warnings=sum(1 for f in prev_findings if f.severity == "WARNING"),
                info=sum(1 for f in prev_findings if f.severity == "INFO")
            )
            formatted = format_scan_report(report)
            formatted.insert(0, "[bold green]>> No files changed since last scan. Showing cached results.[/]")
            formatted.insert(1, "")
            delta_info = {"changed": 0, "new": 0, "deleted": 0, "carried_forward": len(prev_findings)}
            log_dir = save_scan_result(report, formatted, current_fps, delta_info)
            return report, formatted, log_dir

        # Incremental: scan only changed/new files
        delta_files = [fp for fp in all_files if str(fp.relative_to(scan_path)) in changed_files]
        delta_report = await run_code_scan(scan_path, file_filter=delta_files)

        # Carry forward findings from unchanged files
        carried = [f for f in load_previous_findings(prev_data) if f.file_path not in changed_files and f.file_path not in deleted_files]
        all_findings = carried + delta_report.findings

        report = ScanReport(
            scanned_files=len(all_files),
            total_findings=len(all_findings),
            findings=all_findings,
            errors=sum(1 for f in all_findings if f.severity == "ERROR"),
            warnings=sum(1 for f in all_findings if f.severity == "WARNING"),
            info=sum(1 for f in all_findings if f.severity == "INFO")
        )
        formatted = format_scan_report(report)
        formatted.insert(0, f"[bold cyan]>> Incremental Scan[/]: [bold white]{len(changed_files)}[/] changed, [bold white]{len(delta['new'])}[/] new, [bold white]{len(deleted_files)}[/] deleted, [bold white]{len(carried)}[/] carried forward")
        formatted.insert(1, "")
        delta_info = {
            "changed": len(delta["changed"]), "new": len(delta["new"]),
            "deleted": len(delta["deleted"]), "carried_forward": len(carried),
            "changed_files": list(changed_files)
        }
    else:
        # First scan ever — full scan
        report = await run_code_scan(scan_path)
        formatted = format_scan_report(report)
        formatted.insert(0, "[bold cyan]>> Full Scan[/] (first run, no previous baseline)")
        formatted.insert(1, "")

    log_dir = save_scan_result(report, formatted, current_fps, delta_info)
    return report, formatted, log_dir

def format_scan_report(report: ScanReport) -> list[str]:
    # Group findings by category
    categories = {
        "Credentials & Secrets Management": [],
        "Injection Vulnerabilities": [],
        "Transport & Cryptography Security": [],
        "Configuration & Exposure": [],
        "Other Security Findings": []
    }
    
    for f in report.findings:
        rule_id = f.rule_id.upper()
        cat_lower = f.category.lower()
        
        # Determine category group
        if (
            "credential" in cat_lower or 
            rule_id in ("SEC-001", "SEC-002", "SEC-007", "SEC-008", "SEC-009") or
            rule_id.startswith("B105") or rule_id.startswith("B106") or rule_id.startswith("B107")
        ):
            group = "Credentials & Secrets Management"
        elif (
            "injection" in cat_lower or 
            rule_id in ("SEC-003", "SEC-004", "SEC-005", "SEC-014") or
            rule_id.startswith("B6") or rule_id.startswith("B102") or rule_id.startswith("B108")
        ):
            group = "Injection Vulnerabilities"
        elif (
            "transport" in cat_lower or "cryptography" in cat_lower or "crypto" in cat_lower or
            rule_id in ("SEC-006", "SEC-013", "SEC-015") or
            rule_id.startswith("B3") or rule_id.startswith("B5")
        ):
            group = "Transport & Cryptography Security"
        elif (
            "configuration" in cat_lower or "logging" in cat_lower or "config" in cat_lower or
            rule_id in ("SEC-010", "SEC-011", "SEC-012") or
            rule_id.startswith("B1") or rule_id.startswith("B4")
        ):
            group = "Configuration & Exposure"
        else:
            group = "Other Security Findings"
            
        categories[group].append(f)

    lines = [
        f"[bold cyan]Security Scan Report[/]: [bold white]{report.scanned_files}[/] files evaluated",
        f"  [bold red]Errors[/]: {report.errors} | [bold yellow]Warnings[/]: {report.warnings} | [bold cyan]Info[/]: {report.info} | [bold magenta]Total Findings[/]: [bold]{report.total_findings}[/]",
        ""
    ]
    
    has_any = False
    for group, group_findings in categories.items():
        if not group_findings:
            continue
        has_any = True
        lines.append(f"[bold magenta]Category: {group} ({len(group_findings)})[/]")
        for f in group_findings:
            short = f.file_path.split("/")[-1] if "/" in f.file_path else f.file_path
            scolor = "bold red" if f.severity == "ERROR" else "bold yellow" if f.severity == "WARNING" else "cyan"
            lines.append(f"  * [[{scolor}]{f.severity}[/]] [[bold]{f.rule_id}[/]]: [bold white]{f.description}[/] [dim]->[/] [yellow]{short}:{f.line_number}[/]")
            if f.remediation: 
                lines.append(f"    [green]Fix[/]: [white]{f.remediation}[/]")
        lines.append("")
        
    if not has_any:
        lines.append("  [bold green]✓ No security vulnerabilities detected.[/]")
        
    return lines

def _iter_files(root: Path):
    if not root.exists(): return
    for p in root.rglob("*"):
        if p.is_file() and p.suffix in _EXTENSIONS and not any(d in p.parts for d in _SKIP_DIRS):
            yield p
